File: ecommerce_app/lib/models/category.dart
class Category {
  final int id;
  final String name;
  final String? banner;
  final List<Category> children;

  Category({
    required this.id,
    required this.name,
    this.banner,
    required this.children,
  });

  factory Category.fromJson(Map<String, dynamic> json) {
    return Category(
      id: json['id'],
      name: json['name'],
      banner: json['banner'],
      children: (json['children'] as List? ?? [])
          .map((e) => Category.fromJson(e))
          .toList(),
    );
  }
}

File: backend/ecommerce/src/main/java/com/example/ecommerce/model/Category.java
package com.example.ecommerce.model;

import java.util.ArrayList;
import java.util.List;

import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonIgnoreProperties;

import jakarta.persistence.*;
import lombok.Data;

@Data
@Entity
@JsonIgnoreProperties({ "parent", "children" })
@Table(name = "category")
public class Category {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(name = "banner")
    private String banner;

    private String name;

    @ManyToOne
    @JoinColumn(name = "parent_id")
    @JsonIgnore
    private Category parent;

    @OneToMany(mappedBy = "parent", cascade = CascadeType.ALL, fetch = FetchType.EAGER)
    private List<Category> children = new ArrayList<>();

    public boolean isLeaf() {
        return parent != null;
    }
}
